### PR TITLE
warning when light unavilable

### LINKED ISSUE #2762: Lovelace entity not available warnings legibility issue on dark themes (white text on bright yellow bg)
**Home Assistant release with the issue:**

0.87.1

**Browser and Operating System:**

macOS 10.13.6 (17G5019)  + Chrome Version 72.0.3626.109 (Official Build) (64-bit)

**Description of problem:**

For the entity not available alerts, it uses a bright yellow box with the themes "primary-text-color" (i think). This is a big problem for darker themes, since the primary-text-color on those is almost always pure white or something close to it. Bright yellow + pure white text is literally impossible to read.

Lovelace legibility issue

Have reproduced with this issue with 19 different community themes as demonstrated in this brief screen recording: https://gfycat.com/BlindForkedGermanshorthairedpointer  

here is one that you can test it on, used in these screenshots: [midnight theme](https://community.home-assistant.io/t/midnight-theme/28598)

**Additional information:**

After doing more testing, when card is type: entities it displays the text as I would expect, but other card types do not function the same way.

In this photo the "light" card type and "glance" card type are unreadable while the entities card type can be read fine. it's possible that more cards would have this issue but these are the two I noticed in my setup. Config below.

Lovelace legibility issue 2

```
      - type: vertical-stack
        cards:
          - type: horizontal-stack
            cards:
              - type: light
                entity: light.kitchen_cabinets_stove
                icon: mdi:stove
                name: Kitchen Stove Lightstrip
              - type: light
                entity: light.kitchen_cabinets_sink
                icon: mdi:water-pump
                name: Kitchen Sink Lightstrip
```

```
      - type: vertical-stack
        cards:
          - type: glance
            columns: 4
            title: Transmission
            show_name: true
            entities:

              - entity: switch.transmission
                name: Speed Limit
                icon: mdi:traffic-light
                tap_action:
                  action: toggle
```

Lastly I think it might be a neat idea to allow these alert colors to be theme-able in the future.

```
frontend:
  themes:  
    example_theme:
      # Missing Entity Alerts
      missing-entity-background-color: '#000000'
      missing-entity-text-color: '#FFFFFF'
```

### PR DESCRIPTION
**Pull request by iantrich**
Fixes https://github.com/home-assistant/home-assistant-polymer/issues/2762

image

This is the standard we have been using for cards, but I think it's not going to work for lights; personally, out of all my devices, lights are the most often to be unavailable. I think the warning that was there makes more sense (like rows), as it is not as abrasive to a UI.

Thoughts?

### CODE REVIEW

**Comment by balloob** [comment]
I don't think is the right solution. You're changing it from a warning to an error to get around a legibility issue. Why not go to the easy route and just pick a darker yellow? We should probably make a mini custom element that is reusable and has a yellow background, which we can plug in anywhere.

**Comment by iantrich** [comment]
Right. That was my same thought. Error is too much

**Comment by iantrich** [comment]
image

**Comment by iantrich** [comment]
Also took the opportunity to apply decorators and static styles to all entity rows

File: src/panels/lovelace/cards/hui-warn-card.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import {
+  html,
+  LitElement,
+  TemplateResult,
+  property,
+  CSSResult,
+  css,
+} from "lit-element";
+
+export class HuiWarnCard extends LitElement {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-warn-card.ts]: What's different about this one compared to `hui-warn-card` ?

File: src/panels/lovelace/components/hui-warning.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  @property() public warning?: string;
+  @property() public entity?: string;
+
+  public getCardSize(): number {
+    return 1;
+  }
+
+  protected render(): TemplateResult | void {
+    return html`
+      ${this.warning || "Entity not available"}: ${this.entity || ""}
```
Comment (balloob) [inline src/panels/lovelace/components/hui-warning.ts]: I think that we should just render `<slot></slot>` and have the warning be passed in: `<hui-warning>${this.hass.localize('ui.lovelace.warning.entity_not_found', 'entity', this.entity}</hui-warning>`

File: src/panels/lovelace/components/hui-warning.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    `;
+  }
+
+  static get styles(): CSSResult {
+    return css`
+      :host {
+        display: block;
+        color: black;
+        background-color: yellow;
```
Comment (balloob) [inline src/panels/lovelace/components/hui-warning.ts]: This yellow is ugly. Let's use #FCE588

image

File: src/panels/lovelace/components/hui-warning.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  }
+}
+
+declare global {
+  interface HTMLElementTagNameMap {
+    "hui-warning": HuiWarning;
+  }
+}
+
+customElements.define("hui-warning", HuiWarning);
```
Comment (balloob) [inline src/panels/lovelace/components/hui-warning.ts]: Use the decorator `@customElement("hui-warning")` and add on new line above the class definition

```
@customElement("hui-warning")
class HuiWarning extends LitElement {
```

File: src/panels/lovelace/components/hui-warning.ts (hunk 2 above)
Comment (balloob) [reply src/panels/lovelace/components/hui-warning.ts]: You'll have to add the warning string to Lovelace in `translations/en.json` and run `gulp build-translations` to get latest translations build (or restart `script/develop` (slow))

File: src/panels/lovelace/entity-rows/hui-lock-entity-row.ts [hunk 5]
```diff
@@ -56,15 +48,13 @@ class HuiLockEntityRow extends LitElement implements EntityRow {
...
-        mwc-button {
-          color: var(--primary-color);
-          font-weight: 500;
-          margin-right: -0.57em;
-        }
-      </style>
+  static get styles(): CSSResult {
+    return css`
+      mwc-button {
+        color: var(--primary-color);
```
Comment (balloob) [inline src/panels/lovelace/entity-rows/hui-lock-entity-row.ts]: color and font-weight are actually no longer needed, as that's the default. It's left from the conversion from paper-button to mwc-button.

File: src/panels/lovelace/cards/hui-warn-card.ts (hunk 1 above)
Comment (iantrich) [reply src/panels/lovelace/cards/hui-warn-card.ts]: nothing at all. didn't mean to check that it in. my first thought was to create that, and then thought better of it. thanks

File: src/panels/lovelace/components/hui-warning.ts (hunk 4 above)
Comment (iantrich) [reply src/panels/lovelace/components/hui-warning.ts]: decorators are the best!

Comment (iantrich) [reply src/panels/lovelace/components/hui-warning.ts]: I'll probably do a PR today or tomorrow to do this across the board on LitElements, along with property and static css.

**Comment by iantrich** [comment]
Thanks for the great feedback. I believe I got everything all cleaned up. Looking a lot better now, I believe.

**Comment by iantrich** [comment]
image

image

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 6]
```diff
@@ -78,41 +79,53 @@ export class HuiLightCard extends LitElement implements LovelaceCard {
 
     const stateObj = this.hass.states[this._config!.entity] as LightEntity;
 
+    console.log(
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: stale console log.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 7]
```diff
@@ -78,41 +79,53 @@ export class HuiLightCard extends LitElement implements LovelaceCard {
...
+      this.hass.localize(
+        "ui.panel.lovelace.warning.entity_not_found",
+        "entity",
+        this._config.entity
+      )
+    );
+
+    if (!stateObj) {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: This is a great cleanup btw, love it.

File: src/panels/lovelace/components/hui-warning.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  LitElement,
+  TemplateResult,
+  CSSResult,
+  css,
+  customElement,
+} from "lit-element";
+
+@customElement("hui-warning")
+export class HuiWarning extends LitElement {
+  public getCardSize(): number {
```
Comment (balloob) [inline src/panels/lovelace/components/hui-warning.ts]: I don't think that this is needed? We also don't define a `setConfig`.

Comment (iantrich) [reply src/panels/lovelace/components/hui-warning.ts]: Yeah, I questioned that and should have asked. Thanks.

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 6 above)
Comment (iantrich) [reply src/panels/lovelace/cards/hui-light-card.ts]: 🤦‍♂️

File: src/panels/lovelace/entity-rows/hui-script-entity-row.ts [hunk 9]
```diff
@@ -64,15 +62,13 @@ class HuiScriptEntityRow extends LitElement implements EntityRow {
...
-        mwc-button {
-          color: var(--primary-color);
-          font-weight: 500;
-          margin-right: -0.57em;
-        }
-      </style>
+  static get styles(): CSSResult {
+    return css`
+      mwc-button {
+        color: var(--primary-color);
```
Comment (balloob) [inline src/panels/lovelace/entity-rows/hui-script-entity-row.ts]: Out of scope for this PR, but everywhere we can remove the color and font-weight style properties for `mwc-button`, it defaults to the right values.

**Review by balloob — APPROVED** [review]

File: src/panels/lovelace/entity-rows/hui-script-entity-row.ts (hunk 9 above)
Comment (iantrich) [reply src/panels/lovelace/entity-rows/hui-script-entity-row.ts]: https://github.com/home-assistant/home-assistant-polymer/pull/2780

**Pull request merged**